samcli/lib/sync/infra_sync_executor.py

```diff
@@ -319,12 +352,24 @@ def _remove_resource_field(
         if resource_type == AWS_LAMBDA_FUNCTION:
             for field in LAMBDA_FUNCTION_REMOVAL_MAP.get(resource_type, {}).get("Code", []):
                 # We sanitize only if the provided resource is local
+                # Lambda function's Code property accepts dictionary values
                 if (
                     built_resource_dict
+                    and isinstance(built_resource_dict.get("Properties", {}).get("Code"), dict)
                     and is_local_path(built_resource_dict.get("Properties", {}).get("Code", {}).get(field, None))
                 ) or resource_logical_id in linked_resources:
                     resource_dict.get("Properties", {}).get("Code", {}).pop(field, None)
                     processed_logical_id = resource_logical_id
+                # SAM templates also accepts local paths for AWS::Lambda::Function's Code property
+                # Which will be transformed into a dict containing S3Bucket and S3Key after packaging
+                if (
+                    built_resource_dict
+                    and isinstance(built_resource_dict.get("Properties", {}).get("Code"), str)
+                    and is_local_path(built_resource_dict.get("Properties", {}).get("Code"))
+                ):
+                    resource_dict.get("Properties", {}).get("Code", {}).pop("S3Bucket", None)
```

Comment:
nit: Bucket information might stay though 🤔

Reply:
When having local path as "Code", we should remove this field so that we do proper code sync for this scenario. I think even if the bucket is different, we can still leave this for code sync to handle instead of triggering an infra sync

Reply:
Oh got it, yeap makes sense now.